src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,329 @@
+                else if (!isFinalBlock)
+                {
+                    // Don't buffer indefinately
+                    if (span.Length > lengthLimit)
+                    {
+                        throw new InvalidDataException($"Line length limit {lengthLimit} exceeded.");
+                    }
+                    return false;
+                }
+                else
+                {
+                    throw new InvalidDataException("Unexpected end of Stream, the content may have already been read by another component. ");
+                }
+
+                if (line.Length == 0) // an empty line means it's the end of the headers
+                {
+                    return true;
+                }
+
+                if (line.Length > lengthLimit)
+                {
+                    throw new InvalidDataException($"Line length limit {lengthLimit} exceeded.");
+                }
+
+                colon = line.IndexOf(ColonDelimiter);
+
+                if (colon == -1)
+                {
+                    throw new InvalidDataException($"Invalid header line: {line.ToString()}");
+                }
+                else
+                {
+                    key = line.Slice(0, colon);
+                    value = line.Slice(colon + ColonDelimiter.Length);
+                }
+
+                lengthLimit -= line.Length;
+                var decodedKey = GetDecodedString(key);
+                var decodedValue = GetDecodedString(value).Trim();
+
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+            }
+            return false;
+        }
+
+        // For multi-segment parsing of a read only sequence
+        private bool ParseHeadersSlow(
+            ref ReadOnlySequence<byte> buffer,
+            ref KeyValueAccumulator accumulator,
+            ref long headersLength,
+            bool isFinalBlock)
+        {
+            var sequenceReader = new SequenceReader<byte>(buffer);
+            //var position = sequenceReader.Position;
+            //var consumedBytes = default(long);
+
+            while (!sequenceReader.End)
+            {
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> line, CrlfDelimiter))
+                {
+                    if (isFinalBlock)
+                    {
+                        throw new InvalidDataException("Unexpected end of Stream, the content may have already been read by another component. ");
+                    }
+
+                    // Don't buffer indefinately
+                    if (headersLength + sequenceReader.Consumed > HeadersLengthLimit)
+                    {
+                        throw new InvalidDataException($"Line length limit {HeadersLengthLimit - headersLength} exceeded.");
+                    }
+
+                    //TODO: keep the sequence and advance the pipeReader in case the sequence length is shorter than the headers limit.
```

Comment:
Will add one and fix this

Reply:
I am not sure how to create a reliable test for this one - this will only happen if the pipeReader reads a shorter buffer than the actual header size. I don't see a way to control the pipeReader's buffer size other than creating my own pipeReader implementation - which I am not sure if it's the best way.